File: src/subside/plugins/koth/commands/CommandEdit.java
package subside.plugins.koth.commands;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;

import com.sk89q.worldedit.bukkit.WorldEditPlugin;
import com.sk89q.worldedit.bukkit.selections.Selection;

import subside.plugins.koth.areas.Area;
import subside.plugins.koth.areas.Koth;
import subside.plugins.koth.commands.CommandHandler.CommandCategory;
import subside.plugins.koth.exceptions.AreaAlreadyExistException;
import subside.plugins.koth.exceptions.AreaNotExistException;
import subside.plugins.koth.exceptions.CommandMessageException;
import subside.plugins.koth.exceptions.KothNotExistException;
import subside.plugins.koth.modules.Lang;
import subside.plugins.koth.utils.IPerm;
import subside.plugins.koth.utils.MessageBuilder;
import subside.plugins.koth.utils.Perm;
import subside.plugins.koth.utils.Utils;

public class CommandEdit extends AbstractCommand {

    public CommandEdit(CommandCategory category) {
        super(category);
    }

    @Override
    public void run(CommandSender sender, String[] args) {
        if (!(sender instanceof Player)) {
            throw new CommandMessageException(Lang.COMMAND_GLOBAL_ONLYFROMINGAME);
        }

        Player player = (Player) sender;

        if (args.length < 2) {
            Utils.sendMessage(player, true,
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("KoTH editor").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area").commandInfo("Area commands").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot").commandInfo("Loot commands").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> rename <name>").commandInfo("Rename a koth").build());
            return;
        }
        Koth koth = getPlugin().getKothHandler().getKoth(args[0]);
        if (koth == null) {
            throw new KothNotExistException(getPlugin().getKothHandler(), args[0]);
        }

        String[] newArgs = Arrays.copyOfRange(args, 2, args.length);
        if (args[1].equalsIgnoreCase("area")) {
            area(sender, newArgs, koth);
        } else if(args[1].equalsIgnoreCase("loot")){
            loot(sender, newArgs, koth);
        } else if(args[1].equalsIgnoreCase("rename")){
            name(sender, newArgs, koth);
        } else {
            Utils.sendMessage(player, true,
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("KoTH editor").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area").commandInfo("Area commands").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot").commandInfo("Loot commands").build(), 
                    new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> rename <name>").commandInfo("Rename a koth").build());
        }

    }

    private void name(CommandSender sender, String[] args, Koth koth){
        if(args.length < 1){
            throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> rename <name>");
        }
        koth.setName(args[0]);
        getPlugin().getKothHandler().saveKoths();
        throw new CommandMessageException(Lang.COMMAND_EDITOR_NAME_CHANGE);
    }
    
    private void area(CommandSender sender, String[] args, Koth koth) {
        if (args.length > 0) {
            if (args[0].equalsIgnoreCase("create")) {
                Selection selection = ((WorldEditPlugin) getPlugin().getServer().getPluginManager().getPlugin("WorldEdit")).getSelection((Player) sender);
                if (selection != null) {
                    if(args.length < 2){
                        throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> area create <name>");
                    }
                    Location min = selection.getMinimumPoint();
                    Location max = selection.getMaximumPoint();
                    if(koth.getArea(args[1]) != null){
                        throw new AreaAlreadyExistException(args[1]);
                    }
                        
                    Area area = new Area(args[1], min, max);
                    koth.getAreas().add(area);
                    getPlugin().getKothHandler().saveKoths();
                    throw new CommandMessageException(Lang.COMMAND_EDITOR_AREA_ADDED);
                } else {
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_WESELECT);
                }
            } else if (args[0].equalsIgnoreCase("list")) {
                new MessageBuilder(Lang.COMMAND_LISTS_EDITOR_AREA_TITLE).buildAndSend(sender);
                for (Area area : koth.getAreas()) {
                    new MessageBuilder(Lang.COMMAND_LISTS_EDITOR_AREA_ENTRY).area(area).buildAndSend(sender);
                }
                return;
            } else if (args[0].equalsIgnoreCase("edit")) {
                Selection selection = ((WorldEditPlugin) getPlugin().getServer().getPluginManager().getPlugin("WorldEdit")).getSelection((Player) sender);
                if (selection == null) {
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_WESELECT);
                }
                if(args.length < 2){
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> area edit <name>");
                }
                Location min = selection.getMinimumPoint();
                Location max = selection.getMaximumPoint();
                Area area = koth.getArea(args[1]);
                if(area == null){
                    throw new AreaNotExistException(args[1]);
                }
                        
                area.setArea(min, max);
                getPlugin().getKothHandler().saveKoths();
                throw new CommandMessageException(Lang.COMMAND_EDITOR_AREA_EDITED);
            } else if (args[0].equalsIgnoreCase("remove")){
                if(args.length < 2){
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> area edit <name>");
                }
                Area area = koth.getArea(args[1]);
                if(area == null){
                    throw new AreaNotExistException(args[1]);
                }
                koth.getAreas().remove(area);
                getPlugin().getKothHandler().saveKoths();
                throw new CommandMessageException(Lang.COMMAND_EDITOR_AREA_DELETED);
            }
        }
        Utils.sendMessage(sender, true,
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("Area commands").build(), 
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area create <name>").commandInfo("create an area").build(), 
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area edit <area>").commandInfo("re-sets an area").build(), 
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area list").commandInfo("shows the area list").build(),
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area remove <area>").commandInfo("removes an area").build());
    }

    private void loot(CommandSender sender, String[] args, Koth koth){
        Player player = (Player)sender;
        if(args.length > 0){
            if(args[0].equalsIgnoreCase("setpos")){
                Block block;
                try {
                    Method method;
                    try {
                        method = LivingEntity.class.getDeclaredMethod("getTargetBlock", Set.class, int.class);
                    } catch(NoSuchMethodException e){
                        method = LivingEntity.class.getDeclaredMethod("getTargetBlock", HashSet.class, int.class);
                    }
                    block = (Block)method.invoke(player, null, 8);
                } catch(NoSuchMethodException e){
                    getPlugin().getLogger().severe("Cannot find the getTargetBlock function!");
                    e.printStackTrace();
                    return;
                } catch(IllegalAccessException | InvocationTargetException e){
                    getPlugin().getLogger().severe("Cannot access the getTargetBlock function!");
                    e.printStackTrace();
                    return;
                }

                if(block == null){
                    throw new CommandMessageException(Lang.COMMAND_EDITOR_LOOT_SETNOBLOCK);
                }
                koth.setLootPos(block.getLocation());
                getPlugin().getKothHandler().saveKoths();
                throw new CommandMessageException(Lang.COMMAND_EDITOR_LOOT_POSITION_SET);
            } else if(args[0].equalsIgnoreCase("link")){
                if(args.length < 2){
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> loot link <loot>");
                }
                koth.setLoot(args[1]);
                getPlugin().getKothHandler().saveKoths();
                throw new CommandMessageException(Lang.COMMAND_EDITOR_LOOT_LINK);
            } else if(args[0].equalsIgnoreCase("second")){
                if(args.length < 2){
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> loot second (none|north|east|south|west)");
                }
                try {
                    koth.setSecondLootDirection(Koth.LootDirection.valueOf(args[1].toUpperCase()));
                } catch(Exception e){
                    throw new CommandMessageException(Lang.COMMAND_GLOBAL_USAGE[0]+"/koth edit <koth> loot second (none|north|east|south|west)");
                }

                getPlugin().getKothHandler().saveKoths();
                throw new CommandMessageException(Lang.COMMAND_EDITOR_LOOT_SECOND_CHEST);
            }
        }

        Utils.sendMessage(sender, true,
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("loot commands").build(),
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot setpos").commandInfo("sets the position to the block looking at").build(),
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot link <loot>").commandInfo("links a loot chest").build(),
                new MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot second (none|north|east|south|west)").commandInfo("set where second chest will spawn").build()
        );
    }

    @Override
    public IPerm getPermission() {
        return Perm.Admin.EDIT;
    }

    @Override
    public String[] getCommands() {
        return new String[] {
            "edit"
        };
    }
    
    @Override
    public String getUsage() {
        return "/koth edit <koth>";
    }

    @Override
    public String getDescription() {
        return "Edits an existing koth";
    }

}

File: src/subside/plugins/koth/commands/editor/CommandEdit.kt
package subside.plugins.koth.commands.editor

import com.sk89q.worldedit.bukkit.WorldEditPlugin
import org.bukkit.block.Block
import org.bukkit.command.CommandSender
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import subside.plugins.koth.KothManager
import subside.plugins.koth.areas.Area
import subside.plugins.koth.areas.Koth
import subside.plugins.koth.commands.Command
import subside.plugins.koth.modules.Lang
import subside.plugins.koth.utils.IPerm
import subside.plugins.koth.utils.MessageBuilder
import subside.plugins.koth.utils.Perm
import subside.plugins.koth.utils.Utils
import java.util.*

class CommandEdit : Command {
    override fun run(kothManager: KothManager, sender: CommandSender, args: Array<String>) {
        if (sender !is Player) {
            MessageBuilder(Lang.COMMAND_GLOBAL_ONLYFROMINGAME).send(sender)
            return
        }

        if (args.size < 2) {
            sendHelp(sender)
            return
        }

        val koth = kothManager.kothHandler.getKoth(args[0])
        if (koth == null) {
            MessageBuilder(Lang.KOTH_ERROR_NOTEXIST)
                    .koth(kothManager.kothHandler, args[0])
                    .send(sender)
        }

        val newArgs = offsetArgs(args, 1)
        when (args[1].toLowerCase()) {
            "area" -> area(kothManager, sender, newArgs, koth)
            "loot" -> loot(kothManager, sender, newArgs, koth)
            "rename" -> rename(kothManager, sender, newArgs, koth)
            else -> sendHelp(sender)
        }


    }

    private fun rename(kothManager: KothManager, sender: CommandSender, args: Array<String>, koth: Koth) {
        if (args.isEmpty()) {
            MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0].toString() + "/koth edit <koth> rename <name>")
                    .send(sender)
            return
        }

        koth.name = args[0]
        kothManager.kothHandler.saveKoths()

        MessageBuilder(Lang.COMMAND_EDITOR_NAME_CHANGE).send(sender)
    }

    private fun area(kothManager: KothManager, sender: Player, args: Array<String>, koth: Koth) {
        fun showHelp() {
            Utils.sendMessage(sender, true,
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("Area commands").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area create <name>").commandInfo("create an area").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area edit <area>").commandInfo("re-sets an area").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area list").commandInfo("shows the area list").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area remove <area>").commandInfo("removes an area").build())
        }

        if (args.isEmpty()) {
            showHelp()
            return
        }

        val selection = (kothManager.plugin.get()!!.server.pluginManager.getPlugin("WorldEdit") as WorldEditPlugin)
                .getSelection(sender)

        when (args[0].toLowerCase()) {
            "create" -> {
                if (args.size < 2) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0].toString() + "/koth edit <koth> area create <name>")
                            .send(sender)
                    return
                }

                if (selection == null) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_WESELECT).send(sender)
                    return
                }

                if (koth.areas.any { it.name == args[1] }) {
                    MessageBuilder(Lang.AREA_ERROR_ALREADYEXISTS)
                            .area(args[1])
                            .send(sender)
                    return
                }
                val area = Area(args[1], selection.minimumPoint, selection.maximumPoint)
                koth.areas.add(area)
                kothManager.kothHandler.saveKoths()

                MessageBuilder(Lang.COMMAND_EDITOR_AREA_ADDED).send(sender)
            }
            "list" -> {
                MessageBuilder(Lang.COMMAND_LISTS_EDITOR_AREA_TITLE).send(sender)
                koth.areas.forEach {
                    MessageBuilder(Lang.COMMAND_LISTS_EDITOR_AREA_ENTRY).area(it).send(sender)
                }
            }
            "edit" -> {
                if (args.size < 2) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0].toString() + "/koth edit <koth> area edit <name>")
                            .send(sender)
                    return
                }

                if (selection == null) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_WESELECT).send(sender)
                    return
                }

                val area = koth.areas.find { it.name == args[1] }
                if (area == null) {
                    MessageBuilder(Lang.AREA_ERROR_NOTEXIST).area(args[1]).send(sender)
                    return
                }

                area.setArea(selection.minimumPoint, selection.maximumPoint)
                kothManager.kothHandler.saveKoths()

                MessageBuilder(Lang.COMMAND_EDITOR_AREA_EDITED).send(sender)
            }
            "remove" -> {
                if (args.size < 2) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0].toString() + "/koth edit <koth> area edit <name>")
                            .send(sender)
                    return
                }

                val area = koth.areas.find { it.name == args[1] }
                if (area == null) {
                    MessageBuilder(Lang.AREA_ERROR_NOTEXIST).area(args[1]).send(sender)
                    return
                }

                koth.areas.remove(area)
                kothManager.kothHandler.saveKoths()
                MessageBuilder(Lang.COMMAND_EDITOR_AREA_DELETED).send(sender)
            }
            else -> showHelp()
        }
    }

    private fun loot(kothManager: KothManager, sender: Player, args: Array<String>, koth: Koth) {
        fun showHelp() {
            Utils.sendMessage(sender, true,
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("loot commands").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot setpos").commandInfo("sets the position to the block looking at").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot link <loot>").commandInfo("links a loot chest").build(),
                    MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot second (none|north|east|south|west)").commandInfo("set where second chest will spawn").build()
            )
        }

        if (args.isEmpty()) {
            showHelp()
            return
        }

        when (args[0]) {
            "setpos" -> {
                val block: Block? = try {
                    val method = try {
                        LivingEntity::class.java.getDeclaredMethod("getTargetBlock", Set::class.java, Int::class.javaPrimitiveType)
                    } catch (e: NoSuchMethodException) {
                        LivingEntity::class.java.getDeclaredMethod("getTargetBlock", HashSet::class.java, Int::class.javaPrimitiveType)
                    }

                    method.invoke(sender, null, 8) as Block
                } catch (e: Exception) {
                    sender.sendMessage("Cannot use the getTargetBlock function!")
                    return
                }

                if (block == null) {
                    MessageBuilder(Lang.COMMAND_EDITOR_LOOT_SETNOBLOCK).send(sender)
                    return
                }

                koth.loot.lootPos = block.location
                kothManager.kothHandler.saveKoths()
                MessageBuilder(Lang.COMMAND_EDITOR_LOOT_POSITION_SET).send(sender)
            }
            "link" -> {
                if (args.size < 2) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0] + "/koth edit <koth> loot link <loot>")
                            .send(sender)
                    return
                }
                koth.loot.lootName = args[1]
                kothManager.kothHandler.saveKoths()
                MessageBuilder(Lang.COMMAND_EDITOR_LOOT_LINK).send(sender)
            }
            "second" -> {
                if (args.size < 2) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0] + "/koth edit <koth> loot second (none|north|east|south|west)")
                            .send(sender)
                    return
                }

                val direction = Koth.LootDirection.values().find { it.text == args[1].toLowerCase() }
                if (direction == null) {
                    MessageBuilder(Lang.COMMAND_GLOBAL_USAGE[0] + "/koth edit <koth> loot second (none|north|east|south|west)")
                            .send(sender)
                    return
                }
                koth.loot.direction = direction
                kothManager.kothHandler.saveKoths()
                MessageBuilder(Lang.COMMAND_EDITOR_LOOT_SECOND_CHEST).send(sender)
            }
            else -> showHelp()
        }
    }

    fun sendHelp(sender: CommandSender) {
        Utils.sendMessage(sender, true,
                MessageBuilder(Lang.COMMAND_GLOBAL_HELP_TITLE).title("KoTH editor").build(),
                MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> area").commandInfo("Area commands").build(),
                MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> loot").commandInfo("Loot commands").build(),
                MessageBuilder(Lang.COMMAND_GLOBAL_HELP_INFO).command("/koth edit <koth> rename <name>").commandInfo("Rename a koth").build())
    }

    override val permission: IPerm = Perm.Admin.EDIT
    override val commands: Array<String> = arrayOf("edit")
    override val usage: String = "/koth edit <koth>"
    override val description: String = "Edits an existing KoTH"
}